### PR TITLE
Clean the build system and make it work for IBMz

### PR DESCRIPTION
**Pull request by Kachoc**
This PR updates and cleans the build system a bit, and make it work on IBMz architectures.
More specifically, the configure file now computes compatibility flag in a safer and more principled manner.

### CODE REVIEW

**Comment by Kachoc** [comment]
Note that a lot of files changed because I regenerated the extracted files and small changes got propagated, but in practice only 3 files were modified: `Makefile`, `dist/configure` and `dist/Makefile.tmpl`

**Review by msprotz — APPROVED** [review]
I've looked at this closely and it looks good to me. In particular, this moves away from the notion of avx/avx2, and towards a notion of 128-bit vector support / 256-bit vector support.

Some remaining tasks to be addressed as part of a followup:
- fix the C tests that still think in terms of x64 vs arm
- fix the OCaml tests (ditto, do `git grep AVX` in bindings/ocaml)
- fix the EverCrypt layer to also think in terms of 128-bit vector support as opposed to AVX

File: dist/configure [hunk 1]
```diff
@@ -277,58 +284,101 @@ if [[ $target_arch == "aarch64" ]]; then
...
+supports_intrinsics=false
+
 if detect_arm; then
   echo "... detected ARM platform"
-  echo "... $build_target does not support 256-bit arithmetic"
-  echo "BLACKLIST += $(ls *CP256*.c *_256.c *_Vec256.c | xargs)" >> Makefile.config
-  echo "... $build_target does not support legacy vale stubs"
-  echo "BLACKLIST += evercrypt_vale_stubs.c" >> Makefile.config
-  echo "#define Lib_IntVector_Intrinsics_vec256 void *" >> config.h
+  # TODO: poorly worded; rename detect_arm_cc into detect_neon_compiler or
```
Comment (franziskuskiefer) [inline dist/configure:295]: Do you want to do this while you're at it? 🙂

Comment (Kachoc) [reply dist/configure]: Done ;)

**Comment by beurdouche** [comment]
> Hmm... why?

Oh, I see. You added it, then removed it. No worries, go on then... : )

**Comment by franziskuskiefer** [comment]
Looks like something isn't working as intended. (From the Linux CI)

```
2021-02-19T11:00:40.9259123Z [STDERR]In file included from libintvector.h:14,
2021-02-19T11:00:40.9259685Z [STDERR]                 from Hacl_Poly1305_128.h:32,
2021-02-19T11:00:40.9260229Z [STDERR]                 from Hacl_Poly1305_128.c:25:
2021-02-19T11:00:40.9261186Z [STDERR]Hacl_Poly1305_128.c: In function 'Hacl_Impl_Poly1305_Field32xN_128_load_acc2':
2021-02-19T11:00:40.9262447Z [STDERR]/usr/lib/gcc/x86_64-linux-gnu/9/include/smmintrin.h:414:1: error: inlining failed in call to always_inline '_mm_insert_epi64': target specific option mismatch
2021-02-19T11:00:40.9263246Z [STDERR]  414 | _mm_insert_epi64 (__m128i __D, long long __S, const int __N)
2021-02-19T11:00:40.9265281Z [STDOUT]cc -fPIC -I. -I /home/everest/hacl-star/kremlin/include -I /home/everest/hacl-star/kremlin/kremlib/dist/minimal -Wall -Wextra -Werror -std=c11 -Wno-unused-variable -Wno-unknown-warning-option -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-infinite-recursion -g -fwrapv -D_BSD_SOURCE -D_DEFAULT_SOURCE -fPIC -Wno-unused -Wno-parentheses -Wno-deprecated-declarations -Wno-#warnings -Wno-error=cpp -Wno-cpp -g -std=gnu11 -O3    -c -o Hacl_Poly1305_128.o Hacl_Poly1305_128.c
2021-02-19T11:00:40.9266568Z [STDERR]      | ^~~~~~~~~~~~~~~~
2021-02-19T11:00:40.9267365Z [STDOUT]make[2]: Leaving directory '/home/everest/hacl-star/dist/mozilla'
2021-02-19T11:00:40.9267970Z [STDERR]In file included from Hacl_Poly1305_128.h:32,
2021-02-19T11:00:40.9268526Z [STDERR]                 from Hacl_Poly1305_128.c:25:
2021-02-19T11:00:40.9269096Z [STDERR]libintvector.h:138:4: note: called from here
2021-02-19T11:00:40.9269598Z [STDERR]  138 |   (_mm_insert_epi64(x0, x1, x2))
2021-02-19T11:00:40.9270074Z [STDERR]      |   ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-02-19T11:00:40.9270996Z [STDERR]Hacl_Poly1305_128.c:94:8: note: in expansion of macro 'Lib_IntVector_Intrinsics_vec128_insert64'
2021-02-19T11:00:40.9271692Z [STDERR]   94 |   f4 = Lib_IntVector_Intrinsics_vec128_insert64(acc4, (uint64_t)0U, (uint32_t)1U);
2021-02-19T11:00:40.9272291Z [STDERR]      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
Hacl_Blake2b_256.c:38:12: error: AVX vector return without AVX enabled changes the ABI
```

**Comment by victor-dumitrescu** [comment]
Nice work! As @msprotz suggested, I will rebase my opam-specific changes from #405 on top of this.

**Review by victor-dumitrescu — APPROVED** [review]

**Comment by victor-dumitrescu** [comment]
>     * fix the OCaml tests (ditto, do `git grep AVX` in bindings/ocaml)

What would the fix involve? The AVX, ... types here mirror the output of the functions in EverCrypt.AutoConfig2

**Comment by Kachoc** [comment]
> Looks like something isn't working as intended. (From the Linux CI)
> 
> ```
> 2021-02-19T11:00:40.9259123Z [STDERR]In file included from libintvector.h:14,
> 2021-02-19T11:00:40.9259685Z [STDERR]                 from Hacl_Poly1305_128.h:32,
> 2021-02-19T11:00:40.9260229Z [STDERR]                 from Hacl_Poly1305_128.c:25:
> 2021-02-19T11:00:40.9261186Z [STDERR]Hacl_Poly1305_128.c: In function 'Hacl_Impl_Poly1305_Field32xN_128_load_acc2':
> 2021-02-19T11:00:40.9262447Z [STDERR]/usr/lib/gcc/x86_64-linux-gnu/9/include/smmintrin.h:414:1: error: inlining failed in call to always_inline '_mm_insert_epi64': target specific option mismatch
> 2021-02-19T11:00:40.9263246Z [STDERR]  414 | _mm_insert_epi64 (__m128i __D, long long __S, const int __N)
> 2021-02-19T11:00:40.9265281Z [STDOUT]cc -fPIC -I. -I /home/everest/hacl-star/kremlin/include -I /home/everest/hacl-star/kremlin/kremlib/dist/minimal -Wall -Wextra -Werror -std=c11 -Wno-unused-variable -Wno-unknown-warning-option -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-infinite-recursion -g -fwrapv -D_BSD_SOURCE -D_DEFAULT_SOURCE -fPIC -Wno-unused -Wno-parentheses -Wno-deprecated-declarations -Wno-#warnings -Wno-error=cpp -Wno-cpp -g -std=gnu11 -O3    -c -o Hacl_Poly1305_128.o Hacl_Poly1305_128.c
> 2021-02-19T11:00:40.9266568Z [STDERR]      | ^~~~~~~~~~~~~~~~
> 2021-02-19T11:00:40.9267365Z [STDOUT]make[2]: Leaving directory '/home/everest/hacl-star/dist/mozilla'
> 2021-02-19T11:00:40.9267970Z [STDERR]In file included from Hacl_Poly1305_128.h:32,
> 2021-02-19T11:00:40.9268526Z [STDERR]                 from Hacl_Poly1305_128.c:25:
> 2021-02-19T11:00:40.9269096Z [STDERR]libintvector.h:138:4: note: called from here
> 2021-02-19T11:00:40.9269598Z [STDERR]  138 |   (_mm_insert_epi64(x0, x1, x2))
> 2021-02-19T11:00:40.9270074Z [STDERR]      |   ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 2021-02-19T11:00:40.9270996Z [STDERR]Hacl_Poly1305_128.c:94:8: note: in expansion of macro 'Lib_IntVector_Intrinsics_vec128_insert64'
> 2021-02-19T11:00:40.9271692Z [STDERR]   94 |   f4 = Lib_IntVector_Intrinsics_vec128_insert64(acc4, (uint64_t)0U, (uint32_t)1U);
> 2021-02-19T11:00:40.9272291Z [STDERR]      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> ```
> Hacl_Blake2b_256.c:38:12: error: AVX vector return without AVX enabled changes the ABI
> ```

Yes, I'm on it. I think it comes from the fact that the `CFLAGS_128/256` has been moved to `dist/configure`

**Comment by Kachoc** [comment]
> > ```
> > * fix the OCaml tests (ditto, do `git grep AVX` in bindings/ocaml)
> > ```
> 
> What would the fix involve? The AVX, ... types here mirror the output of the functions in EverCrypt.AutoConfig2

Yes, but as far as I understand AVX/AVX2 are sets of instructions that we use to implement the vectorized operations. But we could use other sets of instructions, like those from z15 SIMD (and this PR is a first step towards that). So a more general way of doing would be to rather talk in terms of support for VEC128/VEC256. Of course, we would also need to update EverCrypt.

**Comment by Kachoc** [comment]
Does someone know why there is no `lib_intrinsics.h` in `dist/mozilla` ? This file is present whenever `libintvector.h` is present in all the other dist directories.
If we add it to `dist\mozilla` then I have a way of making the build work. Otherwise I have to make a few modifications, because `dist\mozilla` doesn't behave exactly the same way as the other directories.

**Comment by beurdouche** [comment]
> Does someone know why there is no `lib_intrinsics.h` in `dist/mozilla` ? This file is present whenever `libintvector.h` is present in all the other dist directories.
> If we add it to `dist\mozilla` then I have a way of making the build work. Otherwise I have to make a few modifications, because `dist\mozilla` doesn't behave exactly the same way as the other directories.

These files have a distinct purpose, and in the primitives we currently extract as part of dist, we don't currently use it.
The goal of having specific dist directories is to make sure that there is nothing else than what's strictly needed for each directories.

**Comment by Kachoc** [comment]
I pushed a commit which should fix the problem.
Here is a small explanation (that I included in the commit):
Previously, dist/configure was only copied in dist/gcc-compatible. Now it is copied in more directories, because we now need this script to compute the proper CFLAGS for the vectorized functions. Note that the configure script needs the lib_intrinsics.h file to be present to check if bug 81300 is present or not. If lib_intrinsics.h is absent, it assumes there is no bug. For now it only assumes there is no bug when run from dist/mozilla, which is ok because this assumption was already made before.

**Comment by msprotz** [comment]
The fix looks good to me, thanks @Kachoc! This stuff is always tricky so thanks @franziskuskiefer for spotting this.

@franziskuskiefer I hope that this is in the spirit of your comment here: https://github.com/project-everest/hacl-star/issues/393#issuecomment-773842658 and maybe we can close that other issue?

**Pull request merged**